### PR TITLE
Device Defender changes

### PR DESCRIPTION
**Pull request by TwistedTwigleg**
*Description of changes:*

Adjusts the custom metrics clean up function to clear the memory for the array in custom string-list and ip-list metrics. It also adds support for double numbers and double number lists, allowing for numbers with decimals to be used in the resulting JSON.

*Leaving as draft currently as it may need additional changes*
________

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.*

### CODE REVIEW

File: source/device_defender.c [hunk 1]
```diff
@@ -522,6 +525,24 @@ static int s_get_metric_report_json(
                         cJSON_AddItemToArray(json_list, array_item);
                     }
                     break;
+                case DD_METRIC_NUMBER_DOUBLE:
+                    cJSON_AddNumberToObject(
+                        item, "number", (double)custom_metrics_data[metric_index].data.number_double);
```
Comment (graebm) [inline source/device_defender.c:530]: [trivial] why are you casting this to `(double)`? Isn't it already a `double`?

File: source/device_defender.c [hunk 2]
```diff
@@ -1365,6 +1410,74 @@ int aws_iotdevice_defender_config_register_ip_list_metric(
...
+    const struct aws_byte_cursor *metric_name,
+    aws_iotdevice_defender_get_number_double_fn *supplier,
+    void *userdata) {
+    AWS_PRECONDITION(task_config != NULL);
+    AWS_PRECONDITION(metric_name != NULL);
+    AWS_PRECONDITION(supplier != NULL);
+
+    struct aws_allocator *allocator = task_config->allocator;
+    struct defender_custom_metric *custom_metric = aws_mem_calloc(allocator, 1, sizeof(struct defender_custom_metric));
+    if (custom_metric) {
```
Comment (graebm) [inline source/device_defender.c:1424]: memory allocation can't fail anymore, don't need this `if` check. In the past we didn't guarantee this, so you'll see older code doing the checks. But newer code can be simpler

**Comment by TwistedTwigleg** [comment]
*Will check and fix what is causing the tests to fail. I think it is just something minor due to removing a constant or something*

**Comment by TwistedTwigleg** [comment]
Thanks for the review! I adjusted the code and it's all working again 👍

File: include/aws/iotdevice/device_defender.h [hunk 3]
```diff
@@ -259,13 +259,11 @@ int aws_iotdevice_defender_config_set_callback_userdata(
...
  * \param[in]    userdata       specific callback data for the supplier callback function
- * \returns    AWS_OP_SUCCESS if the custom metric has been associated with the
- *             task configuration successfully
  */
 AWS_IOTDEVICE_API
-int aws_iotdevice_defender_config_register_number_metric(
+void aws_iotdevice_defender_config_register_number_metric(
     struct aws_iotdevice_defender_task_config *task_config,
-    const struct aws_byte_cursor *metric_name,
+    struct aws_byte_cursor *metric_name,
```
Comment (bretambrose) [inline include/aws/iotdevice/device_defender.h:266]: Unsure why const was dropped here.

File: source/device_defender.c [hunk 4]
```diff
@@ -486,10 +486,10 @@ static int s_get_metric_report_json(
                         goto cleanup;
                     }
                     cJSON_AddItemToObject(item, "number_list", json_list);
-                    for (size_t num_index = 0; num_index < list_size; ++num_index) {
-                        int64_t number = 0;
+                    for (size_t num_index = 0; num_index < list_size; num_index++) {
```
Comment (bretambrose) [inline source/device_defender.c:489]: prefer prefix inc/dec if there's no particular reason not to

File: source/device_defender.c [hunk 5]
```diff
@@ -477,7 +477,7 @@ static int s_get_metric_report_json(
 
             switch (custom_metrics_data[metric_index].metric->type) {
                 case DD_METRIC_NUMBER:
-                    cJSON_AddNumberToObject(item, "number", (double)custom_metrics_data[metric_index].data.number);
+                    cJSON_AddNumberToObject(item, "number", custom_metrics_data[metric_index].data.number);
```
Comment (bretambrose) [inline source/device_defender.c:480]: I'm not one to complain about function size very often, but good grief, this is... not great (not your fault obviously, I'm just grumbling at having to try and look over the thing as a whole).

File: source/device_defender.c [hunk 6]
```diff
@@ -643,7 +643,9 @@ static void s_clean_up_custom_metric_data(
                                 &custom_metrics_data[metric_index].data.list, (void *)&string_or_ip_entry, item_index);
                             aws_string_destroy(string_or_ip_entry);
                         }
-                    /* fall through */
+                        aws_array_list_clean_up(
```
Comment (bretambrose) [inline source/device_defender.c:646]: Totally agree with this change.

File: source/device_defender.c (hunk 5 above)
Comment (TwistedTwigleg) [reply source/device_defender.c:480]: Yeah, it's rather large. In the future (maybe with the JSON changes?) I'll take a look at splitting this function down into multiple smaller functions so it is easier to read, debug, and work with.

**Comment by TwistedTwigleg** [comment]
Thanks for the review! I adjusted the code accordingly 👍

**Review by bretambrose — APPROVED** [review]

**Review by bretambrose — APPROVED** [review]

**Comment by TwistedTwigleg** [comment]
Thanks for the reviews! Merging into `main`...

**Pull request merged**